star: Update to 1.7.0

RobK88 opened the pull request:
#### Description

star:  update to version 1.7.0

* Fix broken port
* Reformat Portfile
* Add "makefile 1.0" PortGroup
* Add "conflicts_build 1.0" PortGroup
* Add "codeberg 1.0" PortGroup
* Add codeberg.setup after name
* Update version to 1.7.0-${codeberg.version}
* Add revision number
* Remove "platforms darwin"
* Add myself as a maintainer
* Change the description of the Port
* Modify long_description
* Remove homepage
* Remove master_sites
* Update checksums
* Remove "use_bzip2"
* Delete the patch file - files/patch-RULES__rules.prg.diff
* Add patch block to change all occurrences of gnutar in the man page to sgnutar
* Add patch block to delete "-noclobber" flag in cdrecord/Makefile.dfl
* Add patch block to replace /opt/schily with ${prefix} in the source code files
* Delete worksrcdir
* Add "port_conflict_check" procedure
* Add "port_conflict_check" procedure call to check against build conflicts with cdrtools versions 0 to 3.01_1
* Add "compiler.blacklist-append"
* Remove "use_configure no"
* Add -lintl to configure.ldflags-append
* Remove build dependency - port:gmake
* Add build dependency - port:smake
* Add build dependency - port:gettext
* Change depends_build to depends_build-append
* Add library dependency "port:gettext-runtime" to depends_lib-append
* Add library dependency "port:libiconv" to depends_lib-append
* Change build.cmd to smake
* Delete "build.jobs -1" from the Portfile
* Add "use_parallel_build      no"
* Update note regarding the disabling of parallel builds
* Add INS_BASE, INS_RBASE DEFOSINCDIRS, CC_OPT to build.args-append in the Portfile
* Add INS_RBASE, DEFINSUSR and DEFINSGRP to destroot.destdir in the Portfile
* Replace all system() calls Tcl code and MacPorts macros in the post-destroot code block
* Replace Tcl commands with preferred MacPorts macros wherever possible in the post-destroot code block
* Rename the gnutar binary to sgnutar in destroot
* Delete the rename of the tartest binary in destroot (i.e. keep the tartest name)
* Delete the rename of the tartest man page in destroot (i.e. keep the tartest name)
* Fix post-destroot block to delete unnecessary binaries, manpages and docs
* Delete livecheck related statements
* Add notes to Portfile informing users that the Schily gnutar binary has been renamed sgnutar

CLOSES:  https://trac.macports.org/ticket/49104
<!-- Note: it is best to make pull requests from a branch rather than from master -->
**IMPORTANT - PLEASE READ:**  This PR for `star` will not build if the old version of `cdrtools`, version 3.01, currently in the MacPorts repo has been installed and is active. To address this issue, a procedure called `port_conflict_check` has been added which checks whether `cdrtools`, version 3.01_1 or earlier has been installed. (Thanks to @mascguy for the code).  Once `smake` is built, `cdrtools` can be activated again.

FYI -- The old version of `cdrtools`, version 3.01, in the MacPorts repo installs OLD libschily header files and libraries. It is the existence of these old header files and libraries which prevents this updated port for `star` from building successfully. So the previous version of `cdrtools` must be uninstalled or deactivated first.

A new updated port for `cdrtools` has been submitted which no longer installs the libshily header files and libraries since they are not needed to run the binaries.  Please see https://github.com/macports/macports-ports/pull/18741 

Please see the following related Pull Requests:

(smake)         https://github.com/macports/macports-ports/pull/18655
(cdrtools)      https://github.com/macports/macports-ports/pull/18741

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [x] bugfix
- [ ] enhancement
- [x] security fix

###### Tested on
<!-- Triple-click and copy the next line and paste it into your shell. It will copy your OS and Xcode version to the clipboard. Paste it here replacing this section.
-->
macOS 10.13.6 17G14042 x86_64
Code 10.1 10B61

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? <!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [x] checked your Portfile with `port lint --nitpick`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

RobK88 commented:
Hmm.  Looks like it is failing when linking in the destroot phase.  Strange.  I do not see this happening when running the updated star port on High Sierra.  Any ideas?

catap commented on archivers/star/Portfile:
```diff
@@ -1,51 +1,105 @@
+        rm -rf man3 && \
+        mv man1/gnutar.1 man1/sgnutar.1"
+        
+    file delete -force ${destroot}${prefix}/share/doc/rmt/default-rmt.sample
```
```suggestion
    foreach f {mt smt rmt tar} {
        file delete -force ${destroot}${prefix}/bin/$f
    }

    file rename \
        ${destroot}${prefix}/bin/gnutar \
        ${destroot}${prefix}/bin/sgnutar

    foreach f {man1/mt.1 man1/smt.1 man1/rmt.1 man1/match.1 man5/makerules.5 man5/makefiles.5 man3} {
        file delete -force ${destroot}${prefix}/share/man/$f
    }

    file rename \
        ${destroot}${prefix}/share/man/man1/gnutar.1 \
        ${destroot}${prefix}/share/man/man1/sgnutar.1
```

catap commented on archivers/star/Portfile:
```diff
@@ -1,51 +1,105 @@
 
-use_configure       no
+#
+# Note:  smake no longer builds the star port properly. Must use a modern version of gmake
+#
+depends_build-append     port:gmake
 
-# smake does not support the -j flag
-build.jobs          -1
+build.cmd           ${prefix}/bin/gmake
 
-build.cmd           ${prefix}/bin/smake
+#
+# Must specify INS_BASE and INS_RBASE in both the build and destroot phases
+#
+build.args          GMAKE_NOWARN=true \
+                    INS_BASE=${destroot}${prefix} \
+                    INS_RBASE=${destroot}${prefix}
+                    
+universal_variant   no
 
 destroot.destdir    INS_BASE=${destroot}${prefix} \
+                    INS_RBASE=${destroot}${prefix} \
                     MANDIR=man
+
 post-destroot {
     system "cd ${destroot}${prefix}/bin/ && \
-        rm -f gnutar mt tar && \
-        mv tartest startest"
+        rm -f mt smt rmt tar && \
+        mv gnutar sgnutar"
+        
     system "cd ${destroot}${prefix}/share/man && \
-        rm -rf man1/gnutar.1 man1/match.1 man1/rmt.1 man3 \
+        rm -f man1/mt.1 man1/smt.1 man1/rmt.1 man1/match.1 \
         man5/makerules.5 man5/makefiles.5 && \
-        mv man1/tartest.1 man1/startest.1"
-    file delete -force ${destroot}${prefix}/lib ${destroot}${prefix}/include
+        rm -rf man3 && \
+        mv man1/gnutar.1 man1/sgnutar.1"
```
```suggestion
    foreach f {mt smt rmt tar} {
        file delete -force ${destroot}${prefix}/bin/$f
    }

    file rename \
        ${destroot}${prefix}/bin/gnutar \
        ${destroot}${prefix}/bin/sgnutar

    foreach f {man1/mt.1 man1/smt.1 man1/rmt.1 man1/match.1 man5/makerules.5 man5/makefiles.5 man3} {
        file delete -force ${destroot}${prefix}/share/man/$f
    }

    file rename \
        ${destroot}${prefix}/share/man/man1/gnutar.1 \
        ${destroot}${prefix}/share/man/man1/sgnutar.1
```

catap commented on archivers/star/Portfile:
```diff
@@ -1,51 +1,105 @@
-        mv tartest startest"
+        rm -f mt smt rmt tar && \
+        mv gnutar sgnutar"
+        
     system "cd ${destroot}${prefix}/share/man && \
-        rm -rf man1/gnutar.1 man1/match.1 man1/rmt.1 man3 \
+        rm -f man1/mt.1 man1/smt.1 man1/rmt.1 man1/match.1 \
         man5/makerules.5 man5/makefiles.5 && \
-        mv man1/tartest.1 man1/startest.1"
-    file delete -force ${destroot}${prefix}/lib ${destroot}${prefix}/include
+        rm -rf man3 && \
+        mv man1/gnutar.1 man1/sgnutar.1"
```
```suggestion
    foreach f {mt smt rmt tar} {
        file delete -force ${destroot}${prefix}/bin/$f
    }

    file rename \
        ${destroot}${prefix}/bin/gnutar \
        ${destroot}${prefix}/bin/sgnutar

    foreach f {man1/mt.1 man1/smt.1 man1/rmt.1 man1/match.1 man5/makerules.5 man5/makefiles.5 man3} {
        file delete -force ${destroot}${prefix}/share/man/$f
    }

    file rename \
        ${destroot}${prefix}/share/man/man1/gnutar.1 \
        ${destroot}${prefix}/share/man/man1/sgnutar.1
```

RobK88 replied:
Thanks!  Your suggested code is easier to read.  I will try it.  I cannot commit these changes here.  When I try I see "Applying suggestions on deleted lines is currently not supported.".  Bummer.

catap replied:
@RobK88 you still may change it locally and do a force push to your branch ;)

RobK88 replied:
Applied and pushed all the suggested changes.

catap commented on archivers/star/Portfile:
```diff
@@ -1,51 +1,105 @@
+#
+# Note:  smake no longer builds the star port properly. Must use a modern version of gmake
+#
+depends_build-append     port:gmake
```
```suggestion
depends_build-append \
                    port:gmake
```

RobK88 replied:
I do not see a big change here.  Maybe it is more readable?

catap replied:
Yep, to keep padding on the same level.

RobK88 replied:
Applied the fix and pushed the changes.

catap commented:
@RobK88 it doesn't build on macOS 12:
```
 ==> COMPILING "OBJ/i386-darwin-clang/diff.o"
 ==> COMPILING "OBJ/i386-darwin-clang/restore.o"
 ==> COMPILING "OBJ/i386-darwin-clang/remove.o"
header.c:1351:31: error: use of undeclared identifier 'use_fifo'
                        newvolhdr((char *)NULL, 0, use_fifo);
                                                   ^
1 error generated.
star_fat.c:1794:66: warning: illegal character encoding in string literal [-Winvalid-source-encoding]
                gtprintf("Copyright (C) 1985, 88-90, 92-96, 98, 99, 2000-2019 J<F6>rg Schilling\n");
                                                                               ^~~~
gmake[1]: *** [../RULES/r-gmake.obj:51: OBJ/i386-darwin-clang/header.o] Error 1
gmake[1]: *** Waiting for unfinished jobs....
1 warning generated.
gmake[1]: Leaving directory '/opt/local/var/macports/build/_Users_catap_src_macports-ports_archivers_star/star/work/star-1.6/star'
gmake: *** [RULES/rules1.dir:25: all] Error 2
```

RobK88 commented:
> @RobK88 it doesn't build on macOS 12:
> 
Maybe I should try another compiler?  As I said, the port builds fine on High SIerra presumably using Apple LLVM version 10.0.0 (clang-1000.11.45.5).

catap commented:
> > @RobK88 it doesn't build on macOS 12:
> 
> Maybe I should try another compiler? As I said, the port builds fine on High SIerra presumably using Apple LLVM version 10.0.0 (clang-1000.11.45.5).

Well... CI had failed with the same reason.

RobK88 commented:
FYI -- The makefile for `star` v1.6 is brain dead.  The developer of `star` is also the developer of `smake`.  The latest version of `star` (version 1.6) uses a completely different and very buggy makefile system introduced in the latest version of `smake`.  That is why the latest version of `star` will not build using `smake`.  Unfortunately, the developed of `star` has passed away so the bugs in the `star` makefiles or `smake` are unlikely to get fixed.

Moreover, when building the latest version of `star` using `gmake`, the configure stage is run twice.    And the build phase is a lot slower.  I suspect some code is being compiled twice!  What a waste!

I have hacked the code locally so the latest version of `star` will use the old `smake` makefile system.  With the old makefiles and hacks, `star` only runs the configure stage once and builds a lot faster.   

I am thinking of just hosting the hacked code in a repo on my Github page and changing the port to use that code and `smake` instead.   Maybe that will fix the issue on Big Sur and above.  But I suspect I may need to specify a different (older) `clang` compiler.

RobK88 commented:
Success!!  I needed to specify the compiler in the Portfile.  

On my Mac running High Sierra, I was able to successfully build and install this port using the default Apple `clang-10` compiler that comes with High Sierra.  I was also able to successfully build and install this port using the MacPorts `clang-14` and `clang-15` compilers.  But `clang-15` generates a lot more warning messages such as:

`warning: a function definition without a prototype is deprecated in all versions of C and is not supported in C2x [-Wdeprecated-non-prototype]`

I still have no idea why this port will not build using the default compiler on Big Sur, Monterey and Ventura.

In the end, I decided to specify MacPorts `clang-14` in the `star` Portfile.  Was this a good choice?  Any advice would be most welcome on whether I should specify the latest compiler or an earlier one on the Portfile.

Once the port is finalized, I will rebase my changes and push them.

P.S.  This port is taking much longer to compile that it should!!  As noted earlier, version 1.6 of `star` comes with a  broken makefile system.  `smake` no longer works.  And if one uses `gmake` instead (like in this port),  the configure stage is run twice and it appears that some of the code is even compiled twice.  I have hacked the code to revert back to the original Makefies and rules (found in version 1.5.3 of `star`) but it is not part of this port (yet).  Perhaps I should go this route with this port.  If I do, how do I tell Macports to unzip a file in the files folder and tell it to overwrite files found in the unzipped source files before the port compiles the code?  (Or it may be better if I just host the fixed code on my Github page for others to use, including this port!).

Dave-Allured commented:
@RobK88, there are more recent versions of star and smake in the latest Schily Tools version, 2021-09-18.  https://sourceforge.net/projects/schilytools/

There is about 2-1/2 years of additional development since your version 1.6.  I have not analyzed the changes, but some of the problems you noted may have been solved.  I suggest you consider updating to this latest star version.

catap commented on archivers/star/Portfile:
```diff
@@ -1,51 +1,114 @@
+
+use_bzip2           yes
 
 worksrcdir          star-${version}
```
isn't it default value?

RobK88 replied:
According the the MacPorts Guide, the default value for worksrcdir is ${distname}.  worksrcdir can be used if the extracted source directory has a different name than the distfile.  I will see if I can eliminate it.  

Right now I am focused on updating this PR to `star` version 1.6.1.  So I will need to rewrite the port and probably extract the needed source code from schily tools, modify the RULES (so it van build) and host the source code just for `star`, version 1.6.1 on my GitHib website.

P.S.  I can build `star` v1.6.1 on my Mac.  I did it by extracting the source files needed for 'star` v1.6.1 from the latest Schily Tools tarball and by replacing the RULES directory with  the RULES directory found in `star` version 1.5.3.  

```
Grinchs-MBP:i386-darwin-clang grinch$ ./star -version
star: star 1.6.1 (x86_64-apple-macosx17.7.0) 2021/07/03

Options: acl-POSIX.1e-draft find fflags remote Linux-xattr

Copyright (C) 1985, 88-90, 92-96, 98, 99, 2000-2021 J?rg Schilling
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

It is too bad that the developer did not update his website for `star` himself to the latest version.  Building Schily Tools and then deleting all the binaries you do not need is just too much work for most people who just want to use `star`.

RobK88 replied:
@catap  You were right.  I did not need the `worksrcdir` statement in the Portfile since the default value of `${worksrcdir}` is `${distname}` which equals `${name}-${version}`.   

The `worksrcdir` statement has been deleted from the Portfile.  

Thanks for pointing that out to me.

catap commented on archivers/star/Portfile:
```diff
@@ -1,51 +1,114 @@
 worksrcdir          star-${version}
 
-use_configure       no
+configure.compiler  macports-clang-14
```
It doesn't compile because not all fifo-related stuff is wrapped into `FIFO` macros.

I suggest to fix it on the first place by wrapping it via `#ifdef FIFO` ;)

catap replied:
And I have no idea why switch compiler help

RobK88 commented:
I was able to eliminate the patch file for `rules.prg`.  I have rebased all my changes and did a force push.

Reminder -- I might need to blacklist some compilers like was done in the Portfiles for `smake` and `cdrtools`.  

Next step.  Create a repo for `star`, version 1.6.1 using the latest source for Schily Tools and the RULES found in `star`, version 1.5.3 and then update the Portfile accordingly.  I probably will not get to it for a couple of days.

Dave-Allured commented:
Readme files in the latest Schily Tools version, 2021-09-18, include warnings about not using older **smake** versions.  Macports **smake** is out of date.  My guess is that this is why you had problems with the latest **RULES** package.  I filed an update ticket:  https://trac.macports.org/ticket/67405

I recommend building this **star** version using the latest **smake-bootstrap** and **smake** versions included in Schily Tools 2021-09-18, and also using the latest **RULES** package in the same version.

I tried to build Schily Tools stand-alone on my x86 Mac (Big Sur) with this recipe, and NO source code or makefile changes other than **INS_BASE**.  Compilation seemed to go very well, up to a point.  I saw no sign of any **RULES** problems.  This supports my suggestion to stay original.  My builds failed on a library link stage with an error that I don't know how to fix.

`ld: framework not found IOKit`

RobK88 commented:
@Dave-Allured Many thanks for the post.  I also have problems building the latest version of Schily Tools on my mac running High Sirerra.  It also fails trying to build a binary which is NOT needed for `smake`.

I will try to build just `smake` (by modifying the targets in the Makefile for Schily Tools).  Then I will try to build star 1.6.1 using the `star` source from the Schily Tools source tarball and using the latest `smake` program.  (If I can do that, I can submit a PR request to update `smake` too!)

If I can create a port for `star` version 1.6.1 using just the latest Schily Tools source tarball (and a patch to the Makefile so it only builds `star` and not all the other software (like cdrtools) that would be GREAT!

P.S.  From what I can see, the big difference between the old `smake` (version 1.2.5) and the latest `smake` (version 1.6) is how it handles the rules in the RULES directory.  When trying to build newer Schily software using the old `smake` program, one sees strange errors like:

`smake: No such file or directory. Can not open 'OBJ/i386-darwin-clang/OBJ/i386-darwin-clang/align_test.d'.`

Notice the wrong directory.

What is unknown is what will happen when trying to build older programs using the newer version of `smake`.  Will it cause breakage in other ports using `smake`?  I guess we will find out.

Dave-Allured commented:
@RobK88 , elsewhere in the new readme files are some instructions for building individual components.  I suspect this would be the best approach, but I have not tested.  Once again, I do not think that makefile changes are warranted.

There may be some issues building internal libraries that are needed by "full smake", star, and other components.  My IOKit framework failures both came at link time for a couple of the libraries.  To me this suggests some config issue that I don't understand.

RobK88 commented:
@Dave-Allured I was finally able to build the latest version of Schily Tools on my Mac running High Sierra.  It built all the tools including the latest versions of `smake` and `star`.

```
Grinchs-MBP:i386-darwin-clang grinch$ ./smake -version
Smake release 1.6 2021/08/12 (x86_64-apple-macosx17.7.0) Copyright (C) 1985, 87, 88, 91, 1995-2021 J?rg Schilling
```
```
Grinchs-MBP:i386-darwin-clang grinch$ ./star -version
star: star 1.6.1 (x86_64-apple-macosx17.7.0) 2021/07/03

Options: acl-POSIX.1e-draft find fflags remote Linux-xattr

Copyright (C) 1985, 88-90, 92-96, 98, 99, 2000-2021 J?rg Schilling
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

You cannot build the latest version of the Schily Tools by using the older version of `smake` (version 1.2.5) that is provided by MacPorts.   You must use `gmake`.  But do NOT run just `gmake`.  You must use the `./Gmake` script provided by the source.  That is run the following command in the root of the source tarball:

`./Gmake INS_BASE=/usr/local INS_RBASE=/usr/local`

When I have time I will create a new port for the `Schily Tools` and submit a PR.  But after looking at the README's and examining the Makefiles, it does not look like there is an easy way to tell the `Schily Tools` Makefile system to just create `smake` or just create `star` etc.  Bummer.  I guess one may need to hack the Makefiles to do so.  Or just let MacPorts build all the binaries in the `Schily Tools` tarball and then delete the undesired ones in the destroot stage.

Rob

Dave-Allured commented:
@RobK88 , great progress!  Thanks for the feedback!  I will try Gmake on my Bug Sur Mac, "when I have time".   ;-)

RobK88 commented:
When you run `./Gmake INS_BASE=/usr/local INS_RBASE=/usr/local`, the shells cript first makes `smake` via a BOOTSTRAP.  Then the script makes all the other software programs using the latest version of `smake` that was just built.

If you just want to make the latest version of `smake` and nothing else, you just need to do the following (assuming gnu make or `gmake` is your defualt `make` program):

```
cd 'psmake'
./MAKE-all
```

Once `smake` is built, you will need to install it.  Try `./smake INS_BASE=/usr/locak INS_RBASE=/usr/local install`.  Or just install it manually into `/usr/local/bin`.

Once the latest version of `smake` is built, you can build all the other Schily Tools as follows: 

```
cd .. 
psmake/smake
```

I thought one might be able to build individual binaries found in the Schily Tools source tarball by just changing into the directory of the desired program (e.g. `cd star`) and then running `smake INS_BASE=/usr/local INS_RBASE=/usr/local`.  Nope.  That does not work.  When you do this, it just builds all the dependencies or object files in the `star` directory but it does not do any linking (presumably because it needs other object files created for other binaries like `libfind`  and `rmt`).  To be honest, I did not expect it to work. But it was worth a try!  One really needs to create new Makefiles that includes targets for individual programs.  (Or just create a port that runs the Schily Tools Makefile (which will create all the Schily Tools binaries) and  then the deletes the undesired binaires, man pages etc in the destroot phase).

Well at least it should be easy to upadt the `smake` port.  I will work on that first and come back to this PR.

Rob

RobK88 commented:
> @RobK88 , great progress! Thanks for the feedback! I will try Gmake on my Bug Sur Mac, "when I have time". ;-)

When you give it a try, make sure you have the `gmake` port installed.  Older versions of `gmake` may not work according to the developer.

RobK88 commented:
FYI -- I have submitted a Pull Request for `smake`, update to version 1.6.  
See https://github.com/macports/macports-ports/pull/18655

RobK88 commented:
First attempt at a port for `star` v1.6.1.  The port fails when running via the buildbots at the compile stage.  See buildbot logs for details.  Any ideas?  Maybe try specifying a different compiler?

P.S.  I can compile the latest version of `schilytools` locally and via running this updated port locally on my Mac running High Sierra.  But the port fails in the destroot stage.  See below.   Any ideas on a fix at the destroot stage?

```
.../usr/bin/install -c   OBJ/i386-darwin-clang/cdrecord /opt/local/var/macports/build/_Users_grinch_Development_MacPorts_local-repo_ports_archivers_star/star/work/destroot/opt/local/bin/cdrecord
	==> MAKING "install" ON SUBCOMPONENT "SRCROOT/cdrecord/Makefile.man"
.../usr/bin/install -c   OBJ/i386-darwin-clang/man/cdrecord.1 /opt/local/var/macports/build/_Users_grinch_Development_MacPorts_local-repo_ports_archivers_star/star/work/destroot/opt/local/share/man/man1/cdrecord.1
	==> MAKING "install" ON SUBCOMPONENT "SRCROOT/cdrecord/Makefile.dfl"
	==> MAKING DIRECTORY "/opt/local/var/macports/build/_Users_grinch_Development_MacPorts_local-repo_ports_archivers_star/star/work/destroot/opt/local/etc/default"
.../usr/bin/install -c -noclobber  cdrecord.dfl /opt/local/var/macports/build/_Users_grinch_Development_MacPorts_local-repo_ports_archivers_star/star/work/destroot/opt/local/etc/default/cdrecord
/usr/bin/install: illegal option -- n
usage: install [-bCcpSsv] [-B suffix] [-f flags] [-g group] [-m mode]
               [-o owner] file1 file2
       install [-bCcpSsv] [-B suffix] [-f flags] [-g group] [-m mode]
               [-o owner] file1 ... fileN directory
       install -d [-v] [-g group] [-m mode] [-o owner] directory ...
smake: *** Code 64 (Host is down) from command line for target '/opt/local/var/macports/build/_Users_grinch_Development_MacPorts_local-repo_ports_archivers_star/star/work/destroot/opt/local/etc/default/cdrecord'.
smake: Couldn't make 'install'.
smake: Leaving  '/opt/local/var/macports/build/_Users_grinch_Development_MacPorts_local-repo_ports_archivers_star/star/work/schily-2021-09-18/./psmake/smake'[2] from directory '/opt/local/var/macports/build/_Users_grinch_Development_MacPorts_local-repo_ports_archivers_star/star/work/schily-2021-09-18/cdrecord'
smake: Default commandline target: 'install'
smake: Doing                       exit(1)
smake: *** Code 1 (Operation not permitted) from command line for target 'xinstall'.
smake: The following command caused the error:
if [ ".Makefile.man Makefile.dfl Makefile.doc" != . ] ; then				\
		for XMK_F in Makefile.man Makefile.dfl Makefile.doc XxZzY-- ; do			\
			(						\
			if [ -r ./$XMK_F ] ; then				\
				echo "	==> MAKING \"install\" ON SUBCOMPONENT \"SRCROOT/cdrecord/$XMK_F\""; \
				"/opt/local/var/macports/build/_Users_grinch_Development_MacPorts_local-repo_ports_archivers_star/star/work/schily-2021-09-18/./psmake/smake" -f $XMK_F  XARCH=x86_64-darwin-clang install;	\
			else							\
				if [ $XMK_F != XxZzY-- ]; then		\
					echo "NOTICE: Partial source (SRCROOT/cdrecord/$XMK_F) missing";\
				fi;					\
				:;					\
			fi						\
			);						\
		done							\
	else								\
		: work around v7 shell bug;				\
	fi
smake: Couldn't make 'install'.
smake: Leaving  '/opt/local/var/macports/build/_Users_grinch_Development_MacPorts_local-repo_ports_archivers_star/star/work/schily-2021-09-18/./psmake/smake'[1] from directory '/opt/local/var/macports/build/_Users_grinch_Development_MacPorts_local-repo_ports_archivers_star/star/work/schily-2021-09-18/cdrecord'
smake: Default commandline target: 'install'
smake: Doing                       exit(1)
smake: *** Code 1 (Operation not permitted) from command line for target 'install'.
smake: The following command caused the error:
list="";							\
	if [ -d TARGETS -a -f TARGETS/__slot ]; then				\
		list=`echo TARGETS/[0123456789][0123456789]* |			\
			sed -e 's;TARGETS/[0123456789][0123456789];;g' | sed -e 's;!@!;/;g'`;\
	fi;									\
	for DIR in ${list} ;							\
	do									\
		(								\
		echo "	==> MAKING \"install\" ON SUBDIRECTORY \"SRCROOT/$DIR\"";\
		if [ -d ./$DIR -a -r ./$DIR/Makefile ] ; then			\
			cd ./$DIR && "/opt/local/var/macports/build/_Users_grinch_Development_MacPorts_local-repo_ports_archivers_star/star/work/schily-2021-09-18/./psmake/smake"  XARCH=x86_64-darwin-clang DIRNAME=SRCROOT/$DIR install;		\
		else								\
			echo "NOTICE: Partial source (SRCROOT/$DIR) missing";\
		fi								\
		);								\
	done
smake: Couldn't make 'install'.
gmake: *** [Makefile:13: install] Error 1
Command failed:  cd "/opt/local/var/macports/build/_Users_grinch_Development_MacPorts_local-repo_ports_archivers_star/star/work/schily-2021-09-18" && /opt/local/var/macports/build/_Users_grinch_Development_MacPorts_local-repo_ports_archivers_star/star/work/schily-2021-09-18/Gmake install CC="/opt/local/var/macports/build/_Users_grinch_Development_MacPorts_local-repo_ports_archivers_star/star/work/compwrap/cc/usr/bin/clang" CXX="/opt/local/var/macports/build/_Users_grinch_Development_MacPorts_local-repo_ports_archivers_star/star/work/compwrap/cxx/usr/bin/clang++" OBJC="/opt/local/var/macports/build/_Users_grinch_Development_MacPorts_local-repo_ports_archivers_star/star/work/compwrap/objc/usr/bin/clang" OBJCXX="/opt/local/var/macports/build/_Users_grinch_Development_MacPorts_local-repo_ports_archivers_star/star/work/compwrap/objcxx/usr/bin/clang++" INSTALL="/usr/bin/install -c" INS_BASE="/opt/local/var/macports/build/_Users_grinch_Development_MacPorts_local-repo_ports_archivers_star/star/work/destroot/opt/local" INS_RBASE="/opt/local/var/macports/build/_Users_grinch_Development_MacPorts_local-repo_ports_archivers_star/star/work/destroot/opt/local" MANDIR=man DEFINSUSR=root DEFINSGRP=wheel 
Exit code: 2
Error: Failed to destroot star: command execution failed
Error: See /opt/local/var/macports/logs/_Users_grinch_Development_MacPorts_local-repo_ports_archivers_star/star/main.log for details.
Error: Follow https://guide.macports.org/#project.tickets if you believe there is a bug.
Error: Processing of port star failed
```

RobK88 commented:
P.S.  It appears that the development for Schily Tools (including `star` and `smake` etc) has moved to codeberg.

See
https://codeberg.org/schilytools/schilytools

So there is even a version of `star` that is newer than v1.6.1

RobK88 commented:
FYI -- The most recent version of `star` available at the new Schily Tools repo at codeberg.com is version 1.7.

I can easily compile all the Schily Tools binaries in the Terminal with the `./Gmake` script in the source tarball. I should be able to update the Portfile to use it. But I need to figure out why MacPorts insists on using an old version of git at `/usr/bin/git` and not the most recent version of git at `/opt/local/bin/git`. The most recent version of git installed by MacPorts is in my $PATH according to the `which git`.

```
$ ./star -version
star: star 1.7.0 (x86_64-apple-macosx17.7.0) 2023/01/11

Options: acl-POSIX.1e-draft find fflags remote Linux-xattr

Copyright (C) 1985, 88-90, 92-96, 1998-2021 J?rg Schilling
Copyright (C) 2022-2023 the schilytools team
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

RobK88 commented:
I think I see the problem.  The `./Gmake` script compile the source but is not designed to install the compiled binaries.  I suspect one must use the compiled `smake` program to install the binaries (i.e. `smake INS_BASE=/usr/local INS_RBASE=/usr/local install`).

So I will try adding the following to the Portfile for `star` and will cross my fingers:

`destroot.cmd     OBJ/i386-darwin-clang/smake`

RobK88 commented:
Nope adding `destroot.cmd` did not work.  I still see the following error at the `destroot` stage when installing the port locally on my Mac running High Sierra:

```
==> MAKING DIRECTORY "/opt/local/var/macports/build/_Users_grinch_Development_MacPorts_local-repo_ports_archivers_star/star/work/destroot/opt/local/etc/default"
.../usr/bin/install -c -noclobber  cdrecord.dfl /opt/local/var/macports/build/_Users_grinch_Development_MacPorts_local-repo_ports_archivers_star/star/work/destroot/opt/local/etc/default/cdrecord
/usr/bin/install: illegal option -- n
usage: install [-bCcpSsv] [-B suffix] [-f flags] [-g group] [-m mode]
               [-o owner] file1 file2
       install [-bCcpSsv] [-B suffix] [-f flags] [-g group] [-m mode]
               [-o owner] file1 ... fileN directory
       install -d [-v] [-g group] [-m mode] [-o owner] directory ...
smake: *** Code 64 (Host is down) from command line for target '/opt/local/var/macports/build/_Users_grinch_Development_MacPorts_local-repo_ports_archivers_star/star/work/destroot/opt/local/etc/default/cdrecord'.
smake: Couldn't make 'install'.
smake: Leaving  '/opt/local/var/macports/build/_Users_grinch_Development_MacPorts_local-repo_ports_archivers_star/star/work/schily-2023-04-19/smake/OBJ/i386-darwin-clang/smake'[2] from directory '/opt/local/var/macports/build/_Users_grinch_Development_MacPorts_local-repo_ports_archivers_star/star/work/schily-2023-04-19/cdrecord'
smake: Default commandline target: 'install'
smake: Doing                       exit(1)
smake: *** Code 1 (Operation not permitted) from command line for target 'xinstall'.
smake: The following command caused the error:
if [ ".Makefile.man Makefile.dfl Makefile.doc" != . ] ; then				\
		for XMK_F in Makefile.man Makefile.dfl Makefile.doc XxZzY-- ; do			\
			(						\
			if [ -r ./$XMK_F ] ; then				\
				echo "	==> MAKING \"install\" ON SUBCOMPONENT \"SRCROOT/cdrecord/$XMK_F\""; \
				"/opt/local/var/macports/build/_Users_grinch_Development_MacPorts_local-repo_ports_archivers_star/star/work/schily-2023-04-19/smake/OBJ/i386-darwin-clang/smake" -f $XMK_F  XARCH=x86_64-darwin-clang install;	\
			else							\
				if [ $XMK_F != XxZzY-- ]; then		\
					echo "NOTICE: Partial source (SRCROOT/cdrecord/$XMK_F) missing";\
				fi;					\
				:;					\
			fi						\
			);						\
		done							\
	else								\
		: work around v7 shell bug;				\
	fi
smake: Couldn't make 'install'.
smake: Leaving  '/opt/local/var/macports/build/_Users_grinch_Development_MacPorts_local-repo_ports_archivers_star/star/work/schily-2023-04-19/smake/OBJ/i386-darwin-clang/smake'[1] from directory '/opt/local/var/macports/build/_Users_grinch_Development_MacPorts_local-repo_ports_archivers_star/star/work/schily-2023-04-19/cdrecord'
smake: Default commandline target: 'install'
smake: Doing                       exit(1)
smake: *** Code 1 (Operation not permitted) from command line for target 'install'.
smake: The following command caused the error:
list="";							\
	if [ -d TARGETS -a -f TARGETS/__slot ]; then				\
		list=`echo TARGETS/[0123456789][0123456789]* |			\
			sed -e 's;TARGETS/[0123456789][0123456789];;g' | sed -e 's;!@!;/;g'`;\
	fi;									\
	for DIR in ${list} ;							\
	do									\
		(								\
		echo "	==> MAKING \"install\" ON SUBDIRECTORY \"SRCROOT/$DIR\"";\
		if [ -d ./$DIR -a -r ./$DIR/Makefile ] ; then			\
			cd ./$DIR && "/opt/local/var/macports/build/_Users_grinch_Development_MacPorts_local-repo_ports_archivers_star/star/work/schily-2023-04-19/smake/OBJ/i386-darwin-clang/smake"  XARCH=x86_64-darwin-clang DIRNAME=SRCROOT/$DIR install;		\
		else								\
			echo "NOTICE: Partial source (SRCROOT/$DIR) missing";\
		fi								\
		);								\
	done
smake: Couldn't make 'install'.
Command failed:  cd "/opt/local/var/macports/build/_Users_grinch_Development_MacPorts_local-repo_ports_archivers_star/star/work/schily-2023-04-19" && smake/OBJ/i386-darwin-clang/smake install CC="/opt/local/var/macports/build/_Users_grinch_Development_MacPorts_local-repo_ports_archivers_star/star/work/compwrap/cc/usr/bin/clang" CXX="/opt/local/var/macports/build/_Users_grinch_Development_MacPorts_local-repo_ports_archivers_star/star/work/compwrap/cxx/usr/bin/clang++" OBJC="/opt/local/var/macports/build/_Users_grinch_Development_MacPorts_local-repo_ports_archivers_star/star/work/compwrap/objc/usr/bin/clang" OBJCXX="/opt/local/var/macports/build/_Users_grinch_Development_MacPorts_local-repo_ports_archivers_star/star/work/compwrap/objcxx/usr/bin/clang++" INSTALL="/usr/bin/install -c" INS_BASE="/opt/local/var/macports/build/_Users_grinch_Development_MacPorts_local-repo_ports_archivers_star/star/work/destroot/opt/local" INS_RBASE="/opt/local/var/macports/build/_Users_grinch_Development_MacPorts_local-repo_ports_archivers_star/star/work/destroot/opt/local" MANDIR=man DEFINSUSR=root DEFINSGRP=wheel 
Exit code: 1
Warning: The following existing files were hidden from the build system by trace mode:
  /opt/local/bin/smake
  /var/root/.CFUserTextEncoding
The following file would have been hidden from the build system by trace mode if it existed:
  /usr/bert/share/lib/smake/defaults.smk
Error: Failed to destroot star: command execution failed
Error: See /opt/local/var/macports/logs/_Users_grinch_Development_MacPorts_local-repo_ports_archivers_star/star/main.log for details.
Error: Follow https://guide.macports.org/#project.tickets if you believe there is a bug.
Error: Processing of port star failed
```

RobK88 commented:
I see the same error when the buildbot tries to install the compiled binaries, manpages etc in Ventura.

For some reason, the option `-noclobber` is being used with `../usr/bin/install` which is not allowed.  But the code shown does not include the `-noclobber` option.  Any ideas for a fix or patch?

RobK88 commented:
It looks like the problem in the destroot stage with the "-noclobber" flag originates with the cdrecord/Makefil.dfl.  I may need to patch this file and then cross my fingers again.

```
###########################################################################
# Sample makefile for installing non-localized auxiliary files
###########################################################################
SRCROOT=	..
RULESDIR=	RULES
include		$(SRCROOT)/$(RULESDIR)/rules.top
###########################################################################

INSMODE=	0644
INSDIR=		etc/default
INSBASE=	$(INS_RBASE:%/=%)
INSFLAGS=	-noclobber
TARGET=		cdrecord
AUXSRCEXT=	.dfl
#XMK_FILE=	Makefile.man

###########################################################################
include		$(SRCROOT)/$(RULESDIR)/rules.aux
###########################################################################
```

RobK88 commented:
Success!!  Patching out **-noclobber** in **cdrecord/Makefil.dfl** did the trick, at least when I ran the port locally on my Mac running High Sierra.  Now the **destroot** stage completes without any errors!

But now I see the following warning generated by MacPorts:

```
Warning: violation by /opt/local/ccs
Warning: violation by /opt/local/xpg4
Warning: star violates the layout of the ports-filesystems!
Warning: Please fix or indicate this misbehavior (if it is intended), it will be an error in future releases!
```

Not a problem with this port for `star` since I will delete all these directories and files in **destroot**.  

I also have a lot of other binaries, files and directories for all the other Schily Tools to delete in **destroot** as well.  It will take some time to update the portfile.  When I do that, I will force push all my changes.  

I am now confident that the updated port for `star` will build and install on Big Sur, Monterey and Ventura as well.

P.S. I am not sure what to do with these directories that violate the the layout of the ports-filesystems if I decide to create a port for all the Schily Tools.  Any thoughts?

RobK88 commented:
@catap @Dave-Allured - Thanks for all your help!!  I finally got the updated port for `star` to work!  But before this port (and the `smake`) port can be merged, the port for `cdrtools` needs to be updated.  I will work on that next.

And I suspect the MacPorts developers will not like the use of `git` as a fetch dependency in this port.  If you know of any mirror sites for the most recent version of Schilly Tools, please let me know.

Dave-Allured commented:
@RobK88 -- Great effort.  Thank you.  Here is one mirror with tarballs:
https://mirrors.dotsrc.org/schilytools/

There are several others listed in this release document:
https://codeberg.org/schilytools/schilytools/src/branch/master/AN-2023-04-19

Give these a try, see if any of them avoid that certificate problem.

Dave-Allured commented:
Did you see that mailing list note from Ryan?  MacPorts strongly prefers release tarballs over git fetch.

catap commented:
@RobK88 the same technic which I've used at https://github.com/macports/macports-ports/pull/18655#issuecomment-1553601336 can be repeated here.

RobK88 commented:
@catap @Dave-Allured - Yes, I read the note from Ryan.  I understand why Macports does not like ports using `git` as the fetch method.

But I am using an old Mac running High Sierra.  The version of `git` and curl that comes with High Sierra is unable to connect to and download source from sites that use Let's Encrypt certs.

So I must stick with `git` as the fetch method while developing these ports on this Mac running High Sierra.  I so not see an alternative unless I can find a mirror site that does NOT use Let's Encrypt certs.

Once the port has been finalized, the Portfile can be updated to download tarballs (instead of using git) since the buildbots running newer version of macOS will have no problem downloading source from these websites that use the Let's Encrypt certs.

And once the port is merged, MacPorts will host the source so older Macs will download the source from MacPorts when needed.

P.S. It should also be noted that if the port cannot download the source due to the SSL cert expired issue, livecheck also breaks:

Grinchs-MBP:~ grinch$ port livecheck smake
Error: cannot check if smake was updated (SSL certificate problem: certificate has expired)

So may need to keep the Portfile using git as the fetch method in my local MacPorts repo so I can do the livecheck.  Or uograde my mac.  I plan to upgrade it soon to Monterey.

Dave-Allured commented:
> So I must stick with `git` as the fetch method

All right.  I have to agree.  Thanks for trying the other.  With Codeburg that popular, I am surprised they have not sorted out those certificates by now.

RobK88 commented:
@catap @Dave-Allured - Many thanks for all the suggestions. Since this port is pretty much finalized from my perspective, I will update the port to download tarballs and let the build bots build it. I hope to do that sometime today.

In the meantime, I have been busy updating the `cdrtools` port. As you know, until the port for `cdrtools` has been updated, the Pull Requests (PRs) for `smake` and `star` cannot be merged.

For the Pull Request for `cdrtools`, please see:
https://github.com/macports/macports-ports/pull/18741

RobK88 commented:
@catap @Dave-Allured - I have updated the port using @catap suggested code.  The port no longer uses `git` to fetch the source but now just downloads the source tarball.  Many thanks for all your help!

catap commented on archivers/star/Portfile:
```diff
@@ -1,51 +1,227 @@
 
+livecheck.type          regex
+livecheck.version       ${git_tag}
+livecheck.url           $homepage}/tags.rss
```
```suggestion
livecheck.url           ${homepage}/tags.rss
```

RobK88 commented:
As I feared, installations (or upgrades) of star using this port will fail if an old version of cdrtools (version 3.01 or older) is present.  I added a note warning users of the issue and forced pushed the change.  If there is a better way to deal with this issue, please let me know.

RobK88 commented:
Until I can figure out how to prevent `star` from building only when certain versions of the `cdrtools` port are present, I added the following to the Portfile for `star`:

```
PortGroup                           conflicts_build 1.0
conflicts_build-append              cdrtools
```
I also reformatted the post-destroot code block and replaced the Tcl commands with preferred MacPorts macros wherever possible in the the post-destroot code block.  I do not think replacing `file delete -force` with just `delete` will make much difference.  But apparently, it is much better to use the MacPorts macro `move` instead of the Tcl `file rename` since the MacPorts macro will correctly handle renames that only change the case of a file on a case-insensitive filesystem and the Tcl command may not.

catap commented:
@RobK88 here it is possible to make another approach: close all three your PRs and open a new one which contains 3 commits to update / fix:
1. `cdrtools`;
2. `smake`;
3. `star`.

That way you may achive the right order, and garantie that everything is build with the right (=updated) version :)

RobK88 commented:
@catap - I agree.  I will just use `conflicts_build-append  cdrtools` in all the Portfiles for `smake`, `star` and `cdrtools` for now.  

I asked the great folks on the MacPorts Mailing List whether there is an easy way in a Portfile script to determine whether a port is installed and it version number.  (Maybe via the MacPorts registry?) If I can determine that then I can just invoke `conflicts_build-append  cdrtools` when `cdrtools` v 3.01 or earlier has been installed.

With `conflicts_build-append cdrtools` present in all the PR's now, I do not think the order of PR's merges is that important.  But once the PR for `smake` has been merged, I can update the PR's (or submit new ones) for `star` and `cdrtools` so that these ports have `smake` as a dependency and use`'smake` to build the ports.   (But even if I do not submit updated PR's, the existing PR's will still work fine but they will take a little longer to build from source since they will build a boostrapped version of `smake` first before building the rest of the port)

RobK88 commented:
@capat - Please see the updated description of the PR at the top of this webpage.  I did my best to explain the situation regarding the problems created by the old incompatible versions of `smake` and `cdrtools` in the MacPorts repo.

RobK88 commented:
@catap - Many thanks for the great news. The "codeberg 1.0" PortGroup will defeiitely simplify things. I have rebased my `star` branch to the latest master. I have made the changes to the Portfile, rebased and force pushed.

mascguy commented on archivers/star/Portfile:
```diff
@@ -1,51 +1,204 @@
+                        ${worksrcpath}/ved/ved.h
+}
+
+conflicts_build-append  cdrtools
```
I'm guessing that you'll update this to use the conflict-check proc, as with the other ports?

RobK88 replied:
Yes!  Somehow I forgot to add it in the last push.  

Your procedure and check has been added.   The PR has been rebased and forced pushed.  I am just waiting for the build bots to finish.

mascguy reviewed (approved):
Looks great! If you're done with your changes, I'll go ahead and merge after CI completes.

RobK88 commented:
@mascguy - Many thanks for all your feedback.  I learnt a lot. I am slowly but surely leaning tcl and MacPorts.

P.S.  I only have one PR outstanding now for a simple fix to `diffutils`.  (See https://github.com/macports/macports-ports/pull/18787/files).  But I just submitted that a few days ago.  It can wait unless you want my feedback within the next 7 days.

RobK88 commented:
FYI -- I submitted a PR for the +universal variant for `star`
see
https://github.com/macports/macports-ports/pull/20703
